### ISSUE TITLE
用triplet loss finetune 

### ISSUE THREAD

**Issue by zhaowwenzhong** [deepinsight/insightface#283]
1、CUDA_VISIBLE_DEVICES='0,1,2,3' python -u train_softmax.py --network m1 --loss-type 12 --lr 0.005 --mom 0.0 --per-batch-size 150 --data-dir ../datasets/faces_ms1m_112x112 --pretrained ../model-m1-softmax,50 --prefix ../model-m1-triplet
其实train_softmax.py 没有tripletloss，tripletloss在train.py中。
2、对照和其他的参数设置，作者把初始学习率：0.1->0.005，mom：0.9->0
3、特有参数：triplet-bag-size（defaut=3600），triplet-alpha（default=0.3），triplet-max-ap（default=0）.**在这里，我想问问作者，triplet-bag-size 是否大点好？**（本人设置值=4200）
目前我发现我的训练日志中发现反复调用：call reset（如下），这是否正常？？
**call reset()**
eval 4200 images.. 4200
triplet time stat [0.000102, 14.305246, 1.949595, 0.0, 0.0, 0.0]
found triplets 2071
seq len 6150
**INFO:root:Epoch[0] Batch [10]	Speed: 122.12 samples/sec	lossvalue=0.174548**
INFO:root:Epoch[0] Batch [12]	Speed: 949.29 samples/sec	lossvalue=0.064859
INFO:root:Epoch[0] Batch [14]	Speed: 642.70 samples/sec	lossvalue=0.065269
INFO:root:Epoch[0] Batch [16]	Speed: 652.19 samples/sec	lossvalue=0.059076
INFO:root:Epoch[0] Batch [18]	Speed: 645.43 samples/sec	lossvalue=0.068372
**call reset()**
eval 4200 images.. 8400
triplet time stat [0.000167, 20.816745, 3.800071, 0.0, 0.0, 0.0]
found triplets 2016
seq len 6000
**INFO:root:Epoch[0] Batch [20]	Speed: 115.45 samples/sec	lossvalue=0.173918**
INFO:root:Epoch[0] Batch [22]	Speed: 948.28 samples/sec	lossvalue=0.070200
INFO:root:Epoch[0] Batch [24]	Speed: 639.35 samples/sec	lossvalue=0.068926
INFO:root:Epoch[0] Batch [26]	Speed: 644.25 samples/sec	lossvalue=0.066985
INFO:root:Epoch[0] Batch [28]	Speed: 643.97 samples/sec	lossvalue=0.062959
**call reset()**
eval 4200 images.. 12600
triplet time stat [0.00022899999999999998, 27.907889, 5.54027, 0.0, 0.0, 0.0]
found triplets 1873
seq len 5550
**INFO:root:Epoch[0] Batch [30]	Speed: 124.86 samples/sec	lossvalue=0.185821**
INFO:root:Epoch[0] Batch [32]	Speed: 931.34 samples/sec	lossvalue=0.058671
INFO:root:Epoch[0] Batch [34]	Speed: 648.38 samples/sec	lossvalue=0.066276
INFO:root:Epoch[0] Batch [36]	Speed: 638.66 samples/sec	lossvalue=0.058656
INFO:root:Epoch[0] Batch [38]	Speed: 636.64 samples/sec	lossvalue=0.067309
call reset()
4、我发现每一个bag中的第一个batch的lossvalue总要比后面的大不少，这是否正常？？
**训练具体参数**
Called with argument: Namespace(batch_size=600, beta=1000.0, beta_freeze=0, beta_min=5.0, c2c_mode=-10, c2c_threshold=0.0, center_alpha=0.5, center_scale=0.003, ckpt=1, coco_scale=8.785196082309279, ctx_num=4, cutoff=0, data_dir='/db/train', easy_margin=0, emb_size=256, end_epoch=100000, gamma=0.12, image_channel=3, image_h=112, image_w=96, images_per_identity=5, incay=0.0, logits_verbose=0, loss_type=12, lr=0.005, lr_steps='', margin=4, margin_a=0.0, margin_b=0.0, margin_m=0.5, margin_s=64.0, margin_verbose=0, max_steps=0, mom=0.0, network='r50', noise_sgd=0.0, num_classes=105916, num_layers=50, output_c2c=0, patch='0_0_96_112_0', per_batch_size=150, per_identities=30, power=1.0, prefix='./arcface/7/model-r50', pretrained='/home/isag/insightface/arcface/4/model-r50,11', rand_mirror=1, rescale_threshold=0, scale=0.9993, target='lfw,cfp_fp,agedb_30', train_limit=0, triplet_alpha=0.3, triplet_bag_size=4200, triplet_max_ap=0.0, use_deformable=0, use_val=False, verbose=2000, version_act='prelu', version_input=1, version_output='K', version_se=0, version_unit=3, wd=0.0005)
loading ['/home/isag/insightface/arcface/4/model-r50', '11']

**Comment by zhaowwenzhong** [comment]
@nttstar 为什么要把mom 设置成0呢？而不是0.9？

**Comment by yhw-yhw** [comment]
@zhaowwenzhong why do I use train.py to use triplet loss finetune, I get error.
![image](https://user-images.githubusercontent.com/13210637/43179026-2b19a4e4-9003-11e8-9719-a86323d5bf07.png)

**Comment by zheshipinyinMc** [comment]
@yhw-yhw 你的label解析出来有两个值[value1,value2]，所以报错了，改成label[0]

**Comment by Lucien7786** [comment]
@zhaowwenzhong could you share your LossValueMetric calculate method on eval_metric? 
I am using the 'class LossValueMetric(mx.metric.EvalMetric):' defined in src/train_softmax.py,
and my lossvalue print always be 0.000000. it must be something wrong there, how to fix it?
I am trainning y1 network with arcface loss function on msra + celebrity.

INFO:root:Epoch[0] Batch [20]   Speed: 767.83 samples/sec       acc=0.018814    lossvalue=0.000001
INFO:root:Epoch[0] Batch [40]   Speed: 776.89 samples/sec       acc=0.062388    lossvalue=0.000000
INFO:root:Epoch[0] Batch [60]   Speed: 777.66 samples/sec       acc=0.092467    lossvalue=0.000000
INFO:root:Epoch[0] Batch [80]   Speed: 771.54 samples/sec       acc=0.123493    lossvalue=0.000000
......
INFO:root:Epoch[21] Batch [7480]    Speed: 807.70 samples/sec       acc=0.805078    lossvalue=0.000000
INFO:root:Epoch[21] Batch [7500]    Speed: 799.70 samples/sec       acc=0.804688    lossvalue=0.000000
INFO:root:Epoch[21] Batch [7520]    Speed: 794.87 samples/sec       acc=0.798940    lossvalue=0.000000
INFO:root:Epoch[21] Train-acc=0.807826
INFO:root:Epoch[21] Train-lossvalue=0.000000

**Comment by zhaowwenzhong** [comment]
对的。要改lable。@yhw-yhw

**Comment by zhaowwenzhong** [comment]
@Lucien7786 
class LossValueMetric(mx.metric.EvalMetric):
  def __init__(self):
    self.axis = 1
    super(LossValueMetric, self).__init__(
        'lossvalue', axis=self.axis,
        output_names=None, label_names=None)
    self.losses = []

  def update(self, labels, preds):
    loss = preds[-1].asnumpy()[0]
    self.sum_metric += loss
    self.num_inst += 1.0
    gt_label = preds[-2].asnumpy()

**Comment by Lucien7786** [comment]
@zhaowwenzhong 通过比较工具，看上去仅仅是init函数重新定义了？ 作者用的是__init__函数，我这边使用的是Python 2.7.12，mxnet 1.2.0
/usr/local/lib/python2.7/dist-packages/mxnet/metric.py  Line 68
class EvalMetric(object):
    """Base class for all evaluation metrics.
    def **__init__**(self, name, output_names=None,
                 label_names=None, **kwargs):
        self.name = str(name)
        self.output_names = output_names
        self.label_names = label_names
        self._kwargs = kwargs
        self.reset()

**Comment by nttstar** [comment]
mom还是设置成0.9, 具体参考 train_triplet.py

**Comment by hustzeyu** [comment]
你们是怎么打印triplet的loss的？ @zhaowwenzhong @nttstar @yhw-yhw @zheshipinyinMc @Lucien7786
